PhabricatorBMOAuthProvider.php

```diff
@@ -0,0 +1,400 @@
+    return $this->adapter;
+  }
+
+  public function getConfig() {
+    if (!$this->providerConfig) {
+      $this->providerConfig = $this->getProviderConfig();
+    }
+    return $this->providerConfig;
+  }
+
+  public function configureAdapter($adapter) {
+    $config = $this->getConfig();
+
+    $adapter->setAdapterType(self::ADAPTER_TYPE);
+    $adapter->setAdapterDomain($config->getProperty(self::CONFIG_KEY_BUGZILLA_DOMAIN));
+    $adapter->setAuthenticateURI(
+      id(new PhutilURI(''))
+        ->setProtocol(
+          $config->getProperty(self::CONFIG_KEY_BUGZILLA_PROTOCOL)
+        )
+        ->setDomain($adapter->getAdapterDomain())
+        ->setPath('/auth.cgi')
+        ->setQueryParam('description', $config->getProperty(self::CONFIG_KEY_APP_NAME))
+    );
+  }
+
+  /*
+   *  Methods to guide the user through the auth and login processes
+   */
+
+  protected function renderLoginForm(AphrontRequest $request, $mode) {
+    $adapter = $this->getAdapter();
+
+    $csrf = $this->getAuthCSRFCode($request);
+    $login_uri = $this->getLoginURI();
+    $uri = new PhutilURI($adapter->getAuthenticateURI());
+
+    $uri->setQueryParam('callback',
+      PhabricatorEnv::getURI($login_uri).$csrf.'/');
+
+    $attributes = array('method' => 'GET', 'uri' => (string) $uri);
+    return $this->renderStandardLoginButton($request, $mode, $attributes);
+  }
+
+  public function processLoginRequest(
+    PhabricatorAuthLoginController $controller) {
+
+    $request = $controller->getRequest();
+
+    // When Bugzilla sends us a "back channel" POST during registration...
+    if ($request->isHTTPPost()) {
+      return $this->processLoginRequestBackChannelPost($controller, $request);
+    }
+
+    // The last step in Bugzilla's verification process
+    return $this->processLoginRequestConfirmationGet($controller, $request);
+  }
+
+  private function processLoginRequestBackChannelPost($controller) {
+    $config = $this->getConfig();
+    $account = null;
+    $response = null;
+
+    // Get the `client_api_key` and `client_api_login` from Bugzilla's POST data
+    $post_info = array();
+    try {
+      $post_info = phutil_json_decode(PhabricatorStartup::getRawInput());
+    }
+    catch(Exception $ex) {
+      $this->throwException(
+        pht('Phabricator BMO Authentication failed due to '.
+            'invalid JSON from Bugzilla.')
+      );
+    }
+
+    // Throw exception if either key is not provided by Bugzilla
+    if(!isset($post_info['client_api_key']) || !isset($post_info['client_api_login'])) {
+      $this->throwException(
+        pht('Phabricator BMO Authentication failed due to '.
+            'incomplete JSON from Bugzilla.')
+      );
+    }
+
+    // Generate a transaction code which we'll receive back from Bugzilla
+    // To confirm the API and Client Login information which we saved
+    // In PHP7 this becomes: bin2hex(random_bytes(self::TRANSACTION_CODE_LENGTH))
+    $trans_code = bin2hex(openssl_random_pseudo_bytes(
```

Comment:
I just spotted this in https://github.com/phacility/phabricator/blob/98a62c9e49989cff3c59fab6a05be63a81792901/src/applications/oauthserver/PhabricatorOAuthServer.php:

```php
$code   = Filesystem::readRandomCharacters(32);
```

Should switch to that and then test.